src/arch/x86/x86_64/src/bzimage.rs

```diff
@@ -0,0 +1,293 @@
+            unsafe { transmute(buff) }
+        };
+
+        // first we make sure that the kernel is not too old
+
+        // from https://www.kernel.org/doc/Documentation/x86/boot.txt:
+        // For backwards compatibility, if the setup_sects field contains 0, the
+        // real value is 4.
+        if header.setup_sects == 0 {
+            header.setup_sects = 4;
+        }
+        // check magic numbers
+        if header.boot_flag != MAGIC_AA55 {
+            let i = header.boot_flag;
+            let p = unsafe { &header.boot_flag };
+            write!(w, "boot flag is {:x}, not {:x}, at {:x}, \n\r\n", i, MAGIC_AA55, p).unwrap();
+            return Err("magic number missing: header.boot_flag != 0xaa55");
+        }
+        if header.header != HDRS {
+            let i = header.header;
+            let p = unsafe { &header.header };
+            write!(w, "header.header is {:x}, not {:x} @ {:x}\n\r\n", i, HDRS, p).unwrap();
+            return Err("magic number missing: header.header != 0x53726448");
+        }
+        // only accept version >= 2.12
+        if header.version < 0x020c {
+            let version = header.version;
+            write!(w, "kernel version too old: 0x{:04x}", version).unwrap();
+            return Err("kernel version too old");
+        }
+        // if header.xloadflags | XLF_KERNEL_64 == 0 {
+        //     return Err("kernel has no 64-bit entry point");
+        // }
+        // if header.xloadflags | XLF_CAN_BE_LOADED_ABOVE_4G == 0 {
+        //     return Err("kernel cannot be loaded above 4GiB");
+        // }
+        // if false && header.relocatable_kernel == 0 {
+        //     return Err("kernel is not relocatable");
+        // }
+
+        // calculate offsets
+        // bootparam offset
+        // can we just stick with 90000 for now?
+        //let bp_offset = self.mem_size - PAGE_SIZE;
+        //let cmd_line_offset = bp_offset - PAGE_SIZE;
+
+        let mut bp = BootParams::default();
+        bp.hdr = header;
+
+        // load kernel
+        let mut kernel_offset = (bp.hdr.setup_sects as usize + 1) * 512;
+
+        write!(w, "Kernel offset is {:x}\r\n", kernel_offset).unwrap();
+        // Copy from driver into segment.
+        let mut buf = [0u8; 512];
+        let mut amt: usize = 0;
+        writeln!(w, "Copy {:x} to {:x} for {:x}", kernel_offset + self.rom_base, self.load, 0x300000).unwrap();
+        let mut load = self.load;
+        loop {
+            if amt > 0x300000 {
+                break;
+            }
+            let size = match rom.pread(&mut buf, kernel_offset) {
+                Ok(x) => x,
+                _x => break,
+            };
+            //write!(w, "Copy {:x} to {:x} for {:x}\n", buf.as_ptr() as u64, self.load, size).unwrap();
+            unsafe { copy(buf.as_ptr(), load as *mut u8, size) };
+            kernel_offset += size;
+            load += size;
+            amt += size;
+        }
+
+        // setup e820 tables
```

Comment:
This is a pretty long function, maybe split this chunk out for setting up e820 tables? This way you can also add tests for this functionality

Reply:
yeah that's good idea.